numbat/src/typed_ast.rs

```diff
@@ -122,10 +135,23 @@ pub enum Expression {
         Box<Expression>,
         Type,
     ),
+    /// A special binary operator that has a DateTime as one (or both) of the operands
+    BinaryOperatorForDate(
+        Option<Span>,
+        BinaryOperator,
+        /// LHS must be a DateTime
+        Box<Expression>,
+        /// RHS may be a DateTime, a Time unit, or a String
+        Box<Expression>,
+        Type,
+        /// If true, then this is a date subtration operation
+        bool,
```

Comment:
I'd be in favor of an enum argument inside the AST. Even if it only has two options.